### ISSUE TITLE
New Source: Linkedin Ads

### ISSUE THREAD

**Issue by sherifnada** [airbytehq/airbyte#2589]
## Tell us about the new integration you’d like to have
[Linkedin Ads/Marketing](https://docs.microsoft.com/en-us/linkedin/marketing/)

## Endpoints
todo

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368220419619) by [Unito](https://www.unito.io)

**Comment by cheyura** [comment]
Webhook-based? (no/partially/yes)
No

**Available authentication modes (API key/Oauth/other)**
The LinkedIn API uses OAuth 2.0 for user authorization and API authentication. Applications must be authorized and authenticated before they can fetch data from LinkedIn or get access to member data.



**Creating an account**

 Free LinkedIn Account. LinkedIn Premium Career: $29.99/month. LinkedIn Premium Business: $59.99/month

Before you begin an advertising campaign on LinkedIn, you will need the following:
- LinkedIn Company Page: This is your organization
- Developer App: Create your application
- Approved Marketing Access: Apply for access on the products tab when you create your app:This process is not instantaneous, and may take one or more days for approval 
- OAuth 2.0 access token with rw_ads Permission Scope.

**Has Sandbox environment ?**
Yes. LinkedIn provides the ability to create a test ad account for you to develop, test, and demo your integration. A test ad account provides several benefits, including the ability to create test ads via the API which will never be served and will not require billing.


Follow these steps to set up your test ad account:

Create a test ad account using the /adAccountsV2 endpoint. Include an additional boolean parameter test to indicate this is a test ad account. Note that you can only create one test ad account per developer application.
Using the test ad account, perform other actions like creating and updating ads and creatives as you normally would.
Please note these behaviors in test ad accounts:
All creatives created under a test ad account will never be served in production.
Creatives under the test ad account will be auto-rejected in the review process.
Test campaigns do not require billing details or incur any costs.
No reporting data is available from /adAnalyticsV2 for test campaigns as they do not get served.
Reporting data is not available because impressions do not occur under a test account and no social actions can be measured.
Test accounts do not support Audience segment upload.
Only business accounts support test ad accounts. Test enterprise ad accounts are not available.
You cannot create a test ad account in Campaign Manager. You must create a test ad account using an API endpoint.

**How to populate the account with data?**
It can be done through the UI (Campaign Manager), API, Sandbox API.

**Available streams for sync**
PRODUCTS (Consumer, Marketing, Learning) and SERVICES (Posts, Ad Accounts, Campaign management.)

**Integration supports incremental sync?**
No

**Other information/blockers**
API access tiers: https://docs.microsoft.com/en-us/linkedin/marketing/integrations/marketing-tiers
Before you begin an advertising campaign on LinkedIn, you will need the following:
- LinkedIn Company Page: This is your organization
- Developer App: Create your application
- Approved Marketing Access: Apply for access on the products tab when you create your app:This process is not instantaneous, and may take one or more days for approval 
- OAuth 2.0 access token with rw_ads Permission Scope.

**Comment by cheyura** [comment]
@sherifnada to sign in need a code sent to john@dataline.io

**Comment by cheyura** [comment]
@sherifnada nib budget 10usd per day

**Comment by gr00tie** [comment]
Would love to see this one upvoted. It currently is blocking us from using Airbyte.

**Comment by sherifnada** [comment]
@cheyura let's start a campaign in preparation to implement.

**Comment by bazarnov** [comment]
@sherifnada

While deep scoping of the development process (API Endpoints, flow, access, permissions, authentication and refresh tokens, etc) I would like to admit couple of moments that are important:

- The Customer needs to created the real development application and apply & pass to the 'Marketing Developer Platform' in order to have the access to the `LinkedIn Ads API`, otherwise the connector would not work. This process is described in `Marketing` section [HERE](https://docs.microsoft.com/en-us/linkedin/shared/authentication/getting-access?context=linkedin/context)

- **Auth type:** only `OAuth 2.0` with `redirect_url` required. No other options.
This causes some difficulties with the obtaining access_token for the customer, but not that hard to follow, if we provide proper instructions inside connector documentation. Probably this moment could be less hard if the connector will be available for Cloud Airbyte (enter the minimal credentials needed for the connector and follow the instructions such as: redirect to the LinkedIn > Login > Allow > get the grant auth_code) as described [HERE](https://docs.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow?context=linkedin%2Fcontext&tabs=HTTPS)

- **Access_token:** using the `grant_auth_code` obtained from permanent authentication process , we need to send the POST and obtain the `access_token`, normally the response doesn't have the `refresh_token` inside, to have the `refresh_token`, the customer's development application should have the specific rights for that. granted by LinkedIn Review Team, as documented [HERE](https://docs.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow?context=linkedin%2Fcontext&tabs=HTTPS#refreshing-a-token). 
Otherwise the `access_token` is expired in a 2 month from the date it's obtained, and should be regenerated by repeating the process of grant authentication process, as described [HERE](https://docs.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow?context=linkedin%2Fcontext&tabs=HTTPS)

This connector matches the Cloud Airbyte perfectly (with it's ability to redirections), but the local use-cases of Airbyte could face with problems of access_token expiration after 2 month of usage. 

@sherifnada 
If you have any concerns or a questions about the implementation process, let's discuss them.
